url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/lexer/examples/ExampleLexer.java

```diff
@@ -0,0 +1,94 @@
+package fri.patterns.interpreter.parsergenerator.lexer.examples;
+
+import java.io.*;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.Lexer;
+import fri.patterns.interpreter.parsergenerator.lexer.*;
+import fri.patterns.interpreter.parsergenerator.syntax.Syntax;
+import fri.patterns.interpreter.parsergenerator.syntax.builder.SyntaxSeparation;
+
+/**
+	Sample showing how to build a Lexer from StandardLexerRules building blocks.
+	This Lexer scans C/Java stringdefs, chardefs, integer, float, identifiers, comments,
+	and everything else as "verbatim".
+	
+	@author (c) 2002, Fritz Ritzberger
+*/
+
+public abstract class ExampleLexer
+{
+	/** Test main. */
+	public static void main(String [] args)	{
+		String [][] rules = {
+			{ "token", "stringdef" },
+			{ "token", "chardef" },
+			{ "token", "identifier" },
+			{ "token", "float" },
+			{ "token", "integer" },
+			{ "token", "verbatim" },
+			{ "verbatim", "char",
+						Token.BUTNOT, "space",
+						Token.BUTNOT, "newline",
+						Token.BUTNOT, "letter",
+						Token.BUTNOT, "digit",
+						Token.BUTNOT, "stringdef",
+						Token.BUTNOT, "chardef",
+						Token.BUTNOT, "comment" },
+			{ "ignored", "comment" },
+			{ "ignored", "spaces" },
+			{ "ignored", "newlines" },
+		};
+		Syntax syntax = new Syntax(rules);
+
+		try	{
+			fri.util.TimeStopper ts = new fri.util.TimeStopper();
+
+			SyntaxSeparation separation = new SyntaxSeparation(syntax);
+			LexerBuilder builder = new LexerBuilder(separation.getLexerSyntax(), separation.getIgnoredSymbols());
+			Lexer lexer = builder.getLexer();
+			lexer.setDebug(true);
+			lexer.setTerminals(separation.getTokenSymbols());
+
+			System.err.println("time to build lexer was: "+ts.getInterval());
+
+			InputStream in = ExampleLexer.class.getResourceAsStream("ExampleLexer.java");	//Reader in = new FileReader(args[0]);
+			lexer.setInput(in);
+
+			System.err.println("time to build input was: "+ts.getInterval());
+			
+			Token t;
+			do	{
+				t = lexer.getNextToken(null);
+				//System.err.println(ts.getInterval()+" was time to read token "+t.symbol+" -> "+t.text);
+				
+				if (t.symbol == null)	{
+					lexer.dump(System.err);
+					throw new LexerException("Uninterpretable input!");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [LexerException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [74].  Consider adding a catch clause of this specific exception in the outer try statement.